lib/add_dsi_users_to_spreadsheet.rb

```diff
@@ -0,0 +1,65 @@
+require 'dfe_sign_in_api'
+require 'spreadsheet_writer'
+
+class AddDSIUsersToSpreadsheet
+  def initialize
+    @worksheet = Spreadsheet::Writer.new(DSI_USER_SPREADSHEET_ID, DSI_USER_WORKSHEET_GID, true)
+  end
+
+  def all_service_users
+    @worksheet.clear_all_rows
+    total_page_num = total_page_number
+    (1..total_page_num).each do |page|
+      response = DFESignIn::API.new.users(page: page)
+      raise error_message_for(response) if users_nil_or_empty?(response)
+
+      rows = response_to_rows(response)
+      @worksheet.append_rows(rows)
+
+    rescue StandardError => e
+      Rails.logger.warn("DSI API failed to respond at page #{page} with error: #{e.message}")
+    end
```

Comment:
You're clearing all the rows first so in the event the API errors and fails to return anything would this leave the sheet empty? If so, would it be possible to rewrite this to only clear rows if some data was successfully returned by the API?

Reply:
We resolved this in b71d0d2